flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/TranslationContext.scala

```diff
@@ -59,29 +64,48 @@ object TranslationContext {
       .traitDefs(ConventionTraitDef.INSTANCE)
       .build
 
-    tabNames = Map[AbstractTable, String]()
-
+    tablesRegistry = Map[String, AbstractTable]()
     relBuilder = RelBuilder.create(frameworkConfig)
-
     nameCntr.set(0)
 
   }
 
+  /**
+   * Adds a table to the Calcite schema so it can be used by the Table API
+   */
   def addDataSet(newTable: DataSetTable[_]): String = {
+    val tabName = "DataSetTable_" + nameCntr.getAndIncrement()
+    tables.add(tabName, newTable)
+    tabName
+  }
+
+  /**
+   * Adds a table to the Calcite schema and the tables registry,
+   * so it can be used by both Table API and SQL statements.
+   */
+  @throws[TableException]
+  def registerTable(table: AbstractTable, name: String): Unit = {
 
-    // look up name
-    val tabName = tabNames.get(newTable)
+    val existingTable = tablesRegistry.get(name)
```

Comment:
Should we check that the name does not follow the pattern `DataSetTable_[0-9]+`?